Modules/posixmodule.c

```diff
@@ -10073,9 +10073,29 @@ os__getdiskusage_impl(PyObject *module, path_t *path)
     Py_BEGIN_ALLOW_THREADS
     retval = GetDiskFreeSpaceExW(path->wide, &_, &total, &free);
     Py_END_ALLOW_THREADS
-    if (retval == 0)
+    if (retval == 0) {
+        if (GetLastError() == ERROR_DIRECTORY) {
+            wchar_t *dir_path = NULL;
+
+            dir_path = PyMem_New(wchar_t, path->length + 1);
+            if (dir_path == NULL)
+                return PyErr_NoMemory();
+
+            wcscpy_s(dir_path, path->length + 1, path->wide);
+
+            if (_dirnameW(dir_path) != -1) {
+                Py_BEGIN_ALLOW_THREADS
+                retval = GetDiskFreeSpaceExW(dir_path, &_, &total, &free);
+                Py_END_ALLOW_THREADS
+            }
+            PyMem_Free(dir_path);
+            if (retval)
+                goto success;
+        }
         return PyErr_SetFromWindowsErr(0);
```

Comment:
This is really obscure, but since we're here, `PyMem_Free` may overwrite the error code that is picked up by this function. We should capture `GetLastError()` immediately after the call and then pass it to this function.

Reply:
Got it - I'll make that change.